dask_cuda/device_host_file.py

```diff
@@ -0,0 +1,54 @@
+from zict import Buffer, File, Func
+from zict.common import ZictBase
+from distributed.protocol import deserialize_bytes, serialize_bytelist
+from distributed.worker import weight
+
+try:
+    from cytoolz import partial
+except ImportError:
+    from toolz import partial
+
+import os
+
+
+def _is_device_object(obj):
+    return hasattr(obj, '__cuda_array_interface__')
+
+
+class DeviceHostFile(ZictBase):
```

Comment:
This we can, I have already written something for that, just need to add steps to verify the data at different points. This is in fact the easy test, I'm not sure yet how to do a real system test though, when running Dask scheduler with CUDA worker and client code.

Reply:
We can test this thing without setting up a Dask cluster.

For more integrated tests I recommend reading through this: https://distributed.dask.org/en/latest/develop.html#writing-tests

In particular I recommend the `gen_cluster` style async tests.  This will allow you to easily pause the system and inspect the state of the workers directly.  Maybe something like the following:

```python
from distributed.utils_test import gen_cluster
from dask.distributed import wait
import dask.array as da
import cupy

@gen_cluster(client=True, Worker=CUDAWorker, config={'cuda': ...})
def test_spill_device_to_host(client, scheduler, worker1, worker2):
    rs = da.random.RandomState(RandomState=cupy.random.RandomState)
    x = rs.random(...)

    x = x.persist()
    await wait(x) # wait for this to finish
    
    assert worker1.data.device....  # check things about the data 
    assert worker1.data.host....

    future = client.compute(x.sum())  # make sure that things can come back
    result = await future
```

Beware that using async tests can be a little tricky (you need to put `await`s in the right places and be careful not to use blocking API), but the ability to directly inspect state is pretty useful.

Reply:
@pentschev my recommendation for a first pass at a system test is above ^^

That should start everything up in a single thread, so it should be easy to test and inspect.

Reply:
Great, thanks for this suggestion. Sorry for missing this very useful comment earlier!